### ISSUE TITLE
Use with mocha and babel-plugin-rewire

### ISSUE THREAD

**Issue by petrgazarov** [TypeStrong/ts-node#387]
Hi,

I have the following setup:

```
  "dependencies": {
    "typescript": "^2.4.1"
  },
  "devDependencies": {
    "babel-core": "^6.25.0",
    "babel-plugin-rewire": "^1.1.0",
    "babel-preset-es2015": "^6.24.1",
    "mocha": "^3.4.2",
    "ts-node": "^3.2.0"
  },
  "babel": {
    "presets": [
      "es2015"
    ],
    "plugins": [
      "babel-plugin-rewire"
    ]
  }
```

And I'm trying to run a `mocha` test, and getting `TypeError: _get__(...) is not a function` for `__Rewire__`.

Normally you could say

```
import { something, __Rewire__, __ResetDependency__ } from '../';

__Rewire__('variable', newValue);
```

But after ts-node compilation, `__Rewire__` import is undefined. I don't know the root cause here, but worth noting [this TypeScript issue](https://github.com/Microsoft/TypeScript/issues/6264)

Here is my tsconfig.json:
```
{
  "compilerOptions": {
    "moduleResolution": "node",
    "sourceMap": true,
    "module": "commonJS",
    "target": "es2015",
    "allowJs": true,
    "baseUrl": "./"
  }
}
```

And my script:

`    "test": "./node_modules/mocha/bin/mocha --require ts-node/register --require babel-register ./test/index.test.js"
`

Here is a repo with reproducible issue: https://github.com/petrgazarov/typescript-rewire-babel-plugin-example

Would be awesome to get some help here, as Mocha + babel-plugin-rewire usage is widespread in our application.

**Comment by blakeembrey** [comment]
Have you tried checking the compilation output with `tsc` and compared to `tsc`? That would tell us what the issue is. I'd guess it's an issue with `rewire` though - check out the limitations: https://github.com/jhnns/rewire#limitations.

**Comment by petrgazarov** [comment]
Hm, when I use `tsc` to compile, the resulting js file is working fine.

ts file:
```
let message: string = 'foo';

const something = (): void => {
  console.log(message);
}

export { something };
```

becomes js:
```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
let message = 'foo';
const something = () => {
    console.log(message);
};
exports.something = something;
//# sourceMappingURL=index.js.map
```

When mocha runs with `babel-plugin-rewire` against the resulting js file, it works well.

So I think there is a difference between `tsc` compilation and `ts-node` compilation.
Is there a way to check `ts-node` compilation result? 

`ts-node src/index.ts` just runs the code.

**Comment by blakeembrey** [comment]
You can try finding it in the cache directory. It should be the same output. The only other thing I'd suggest is whether `babel-plugin-rewire` is maybe only running on `.js` and not `.ts` then.

**Comment by petrgazarov** [comment]
Confirmed that `babel-plugin-rewire` was not running on `.ts` files in my case.

When using `--require babel-register`, there is currently no way to specify extensions, because extensions [are not managed from `.babelrc` file](https://github.com/babel/babel/issues/3741).

Instead, the only setup that worked for me was to use webpack compilation via [mocha-webpack](https://github.com/zinserjan/mocha-webpack). I used nearly identical webpack config for development and test environments, and thus when mocha ran, the code was precompiled with webpack, which in my case has `awesome-typescript-loader` and `babel`.

I did have to tell `awesome-typescript-loader` to `useBabel: true`, otherwise it renamed import paths before it got to babel.
